File: Day149_2025/Q149_GFG.java
class Q149_GFG {
    private void helper(Node root, int sum, int len, int[] maxLen, int[] maxSum) {
        if (root == null) {
            if (len > maxLen[0]) {
                maxLen[0] = len;
                maxSum[0] = sum;
            } else if (len == maxLen[0] && sum > maxSum[0]) {
                maxSum[0] = sum;
            }
            return;
        }

        helper(root.left, sum + root.data, len + 1, maxLen, maxSum);
        helper(root.right, sum + root.data, len + 1, maxLen, maxSum);
    }

    public int sumOfLongRootToLeafPath(Node root) {
        if (root == null) return 0;
        int[] maxSum = {Integer.MIN_VALUE};
        int[] maxLen = {0};
        helper(root, 0, 0, maxLen, maxSum);
        return maxSum[0];
    }
}

File: Day149_2025/Q149_GFG.cpp
/*
class Node {
  public:
    int data;
    Node *left;
    Node *right;

    Node(int x) {
        data = x;
        left = NULL;
        right = NULL;
    }
}; */

class Solution {
    void helper(Node *root, int sum, int len, int &maxLen, int &maxSum)
    {
        if (!root)
        {
            if (maxLen < len)
            {
                maxLen = len;
                maxSum = sum;
            }
            else if (maxLen == len && maxSum < sum)
                maxSum = sum;
            return;
        }
        helper(root->left, sum + root->data, len + 1, maxLen, maxSum);
        helper(root->right, sum + root->data, len + 1, maxLen, maxSum);
    }
  public:
    int sumOfLongRootToLeafPath(Node *root) {
        // Q. Sum of nodes on the longest path
        // code here
        if (!root)
            return 0;
        int maxSum = INT_MIN, maxLen = 0; 
        helper(root, 0, 0, maxLen, maxSum);
        return maxSum;
    }
};